File: python/src/agent_framework/core/memory_context_engine.py
from __future__ import annotations

from agent_framework.core.contracts import ContextEngine, MemoryProvider
from agent_framework.core.domain import AgentMode, Message


class MemoryAugmentedContextEngine(ContextEngine):
    def __init__(self, inner: ContextEngine, memory: MemoryProvider, limit: int = 5) -> None:
        self._inner = inner
        self._memory = memory
        self._limit = limit

    async def build(self, session_id: str, mode: AgentMode, history: list[Message]) -> tuple[str, list[Message]]:
        system_prompt, messages = await self._inner.build(session_id, mode, history)
        query = _extract_last_user_text(history)
        if not query:
            return system_prompt, messages

        entries = await self._memory.search(query, self._limit)
        if not entries:
            return system_prompt, messages

        memory_section = "\n".join(f"- {entry['key']}: {entry['value']}" for entry in entries)
        augmented = f"{system_prompt}\n\n## Memory\n{memory_section}"
        filtered = [message for message in messages if message.role != "system"]
        return augmented, [Message(role="system", content=augmented), *filtered]


def _extract_last_user_text(history: list[Message]) -> str:
    for message in reversed(history):
        if message.role == "user" and isinstance(message.content, str) and message.content.strip():
            return message.content.strip()
    return ""

File: typescript/packages/core/src/memory-context-engine.ts
import type {
  ContextBuildRequest,
  ContextBuildResult,
  ContextEngine,
  MemoryProvider,
} from './types/contracts.js';
import type { Message } from './types/domain.js';

export interface MemoryAugmentedContextEngineOptions {
  inner: ContextEngine;
  memory: MemoryProvider;
  limit?: number;
}

export class MemoryAugmentedContextEngine implements ContextEngine {
  private readonly limit: number;

  constructor(private readonly options: MemoryAugmentedContextEngineOptions) {
    this.limit = options.limit ?? 5;
  }

  async build(request: ContextBuildRequest): Promise<ContextBuildResult> {
    const result = await this.options.inner.build(request);
    const query = extractLastUserText(request.history);
    if (!query) {
      return result;
    }

    const entries = await this.options.memory.search(query, this.limit);
    if (entries.length === 0) {
      return result;
    }

    const memorySection = entries
      .map((entry) => `- ${entry.key}: ${entry.value}`)
      .join('\n');
    const systemPrompt = `${result.systemPrompt}\n\n## Memory\n${memorySection}`;
    const messages = prependSystem(result.messages, systemPrompt);
    return { systemPrompt, messages };
  }
}

function extractLastUserText(history: Message[]): string {
  for (let index = history.length - 1; index >= 0; index -= 1) {
    const message = history[index];
    if (message?.role === 'user' && typeof message.content === 'string' && message.content.trim()) {
      return message.content.trim();
    }
  }
  return '';
}

function prependSystem(history: Message[], systemPrompt: string): Message[] {
  const withoutSystem = history.filter((message) => message.role !== 'system');
  return [{ role: 'system', content: systemPrompt }, ...withoutSystem];
}
